### DISCUSSION TITLE
How to kill a spark query operation?

### DISCUSSION THREAD

**Discussion by MilkWangStudio** [apache/kyuubi#3822] [Q&A] [answered]
I shoud devleop a sql editor. it connect spark wtih kyuubi USER LEVEL, Use hive-jdbc to send query request and wait response.
How can i get the query job process and kill it when user click KILL button on page.

I tried use hive-jdbc's statement.close function, but kyuubi front server response "Invalid OperationHandle". It's like not implement this feature?

kyuubi version is 1.6.1. Want some helps, Thanks!

**Comment by pan3793** [comment, marked as answer]
Do you mean `hive-jdbc`? It's not part of Kyuubi.

**Reply by MilkWangStudio** [reply]
Yes, But the jdbc connect kyuubi front server use thirft api , I see the kyuubi doc: 
https://kyuubi.apache.org/docs/r1.6.1-incubating/client/jdbc/hive_jdbc.html

it said fully compatible with Hive JDBC.

**Reply by MilkWangStudio** [reply]
Maybe there are other ways to implement this function?
I just want to kill a running spark job, like the Ide DataGrip Stop Button.

**Reply by pan3793** [reply]
Yes, Hive JDBC is supported, but for such client usage and questions, please report to Hive community.

**Reply by pan3793** [reply]
I didn't take a deep look at the Hive code, Kyuubi has a modified Kyuubi Hive JDBC Driver, my following answer is based on this one.

JDBC is absolutely a sync API, and is not thread-safe, any call across threads of stmt is not officially supported. But the underlying thrift API is async, directly calling the async thrift API is straightforward.

`KyuubiStatement#executeAsync` could be used as well.

**Reply by MilkWangStudio** [reply]
Ok, Thanks~ I will report it to Hive Community.

For this Discussion, I want some idea for two things.
1.When running a spark query, How to get this progress of timeline? I should display on the page.
2.How to kill the query job, when user click Kill btn on my page.

I search some keyworks on kyubbi doc, just find a RestfullAPI、JDBC maybe can do this.
Thanks！

**Reply by pan3793** [reply]
see https://github.com/apache/incubator-kyuubi/pull/2529 for how to get query progress

**Reply by pan3793** [reply]
Specifically to your original question

>  kyuubi front server response "Invalid OperationHandle".

The description is too rough.

- Have you searched the log about the op uuid across the services log to see what happened on such op?
- any stack trace and context?

**Reply by MilkWangStudio** [reply]
I just debuging the exception stack. And find it's a mult-thread bug. Like you said:
> JDBC is absolutely a sync API, and is not thread-safe

When the query thread is running ,  the other thread call the statement.close function, it will make kyuubi close the query job.
But the client's query thread always fetch result from kyuubi server. then make this exception.

It's **hive-jdbc**'s question. 

And thanks your reply for the query progress ^_^.
